git-worktree/src/lib.rs

```diff
@@ -1 +1,111 @@
 #![forbid(unsafe_code, rust_2018_idioms)]
+//! Git Worktree
+
+use git_hash::oid;
+use git_index::{
+    entry::{Flags, Mode},
+    State,
+};
+use git_object::bstr::ByteSlice;
+use git_object::Data;
+use quick_error::quick_error;
+use std::convert::TryFrom;
+use std::fs;
+use std::fs::create_dir_all;
+use std::path::Path;
+
+#[cfg(unix)]
+use std::os::unix::fs::PermissionsExt;
+
+quick_error! {
+    #[derive(Debug)]
+    pub enum Error {
+        Utf8Error(err: git_object::bstr::Utf8Error) { from() }
+        TimeError(err: std::time::SystemTimeError) { from() }
+        ToU32Error(err: std::num::TryFromIntError) { from() }
+        IoError(err: std::io::Error) { from() }
+    }
+}
+
+/// Copy index to `path`
+pub fn copy_index<P, Find>(state: &mut State, path: P, mut find: Find, opts: Options) -> Result<(), Error>
+where
+    P: AsRef<Path>,
+    Find: for<'a> FnMut(&oid, &'a mut Vec<u8>) -> Option<Data<'a>>,
```

Comment:
Let's let the return object be a `git_object::BlobRef` (_fully qualified_) instead of bare data. This forces an additional type check which should never fail.